File: lib/lavash/graph.ex
defmodule Lavash.Graph do
  @moduledoc """
  Pure graph algorithms for dependency graphs.

  All functions operate on a deps map: `%{name => [deps]}` where each key
  is a derived field and the value is the list of fields it depends on.

  Used by:
  - `Rx.Graph` — runtime reactive graph
  - `ColocatedTransformer` — compile-time JS metadata generation
  - `GraphMacro` — defgraph compile-time JS generation
  - JS hook (`lavash_optimistic.js`) — consumes the serialized output client-side
  """

  @doc """
  Topological sort via Kahn's algorithm.

  Returns derived field names in dependency order (leaves first).
  State fields (names not in the deps map keys) are ignored.

      iex> Lavash.Graph.topo_sort(%{doubled: [:count], quad: [:doubled]})
      [:doubled, :quad]
  """
  def topo_sort(deps_map) when is_map(deps_map) do
    names = Map.keys(deps_map)
    derive_names = MapSet.new(names)

    in_degree =
      Map.new(names, fn name ->
        count = deps_map[name] |> Enum.count(&MapSet.member?(derive_names, &1))
        {name, count}
      end)

    queue = for {name, 0} <- in_degree, do: name
    kahn(queue, in_degree, deps_map, derive_names, [])
  end

  @doc """
  Reverse dependency index.

  Returns `%{field => [derived fields that depend on it]}`.

      iex> Lavash.Graph.build_dependents(%{doubled: [:count], quad: [:doubled]})
      %{count: [:doubled], doubled: [:quad]}
  """
  def build_dependents(deps_map) when is_map(deps_map) do
    Enum.reduce(deps_map, %{}, fn {name, dep_list}, acc ->
      Enum.reduce(dep_list, acc, fn dep, inner_acc ->
        Map.update(inner_acc, dep, [name], &[name | &1])
      end)
    end)
  end

  @doc """
  BFS to find all fields transitively affected by `changed_fields`.

  Uses the `dependents` map (output of `build_dependents/1`).

      iex> dependents = %{count: [:doubled], doubled: [:quad]}
      iex> Lavash.Graph.transitive_dependents(dependents, [:count])
      MapSet.new([:doubled, :quad])
  """
  def transitive_dependents(dependents, changed_fields) do
    do_transitive(dependents, changed_fields, MapSet.new())
  end

  defp do_transitive(_dependents, [], visited), do: visited

  defp do_transitive(dependents, [field | rest], visited) do
    direct = Map.get(dependents, field, [])
    new = Enum.reject(direct, &MapSet.member?(visited, &1))
    visited = Enum.reduce(new, visited, &MapSet.put(&2, &1))
    do_transitive(dependents, rest ++ new, visited)
  end

  # Kahn's algorithm internals
  defp kahn([], _in_degree, _deps_map, _derive_names, result), do: Enum.reverse(result)

  defp kahn([node | rest], in_degree, deps_map, derive_names, result) do
    dependents =
      for {name, dep_list} <- deps_map,
          node in dep_list,
          MapSet.member?(derive_names, name),
          do: name

    {in_degree, new_ready} =
      Enum.reduce(dependents, {in_degree, []}, fn dep, {deg, ready} ->
        new_deg = Map.update!(deg, dep, &(&1 - 1))
        if new_deg[dep] == 0, do: {new_deg, [dep | ready]}, else: {new_deg, ready}
      end)

    kahn(rest ++ new_ready, in_degree, deps_map, derive_names, [node | result])
  end
end

File: priv/static/graph.js
/**
 * Pure graph algorithms for dependency graphs.
 *
 * Mirrors Lavash.Graph on the Elixir side.
 * Used by LavashOptimistic for client-side reactive recomputation.
 */

/**
 * BFS to find all fields transitively affected by `changedFields`.
 *
 * @param {Object} dependents - reverse index: {field: [dependents]}
 * @param {string[]} changedFields - fields that changed
 * @returns {Set<string>} all transitively affected derived fields
 */
export function findAffected(dependents, changedFields) {
  const affected = new Set();
  const queue = [...changedFields];

  while (queue.length > 0) {
    const field = queue.shift();
    const direct = dependents[field] || [];
    for (const dep of direct) {
      if (!affected.has(dep)) {
        affected.add(dep);
        queue.push(dep);
      }
    }
  }

  return affected;
}

/**
 * Recompute derived fields in topological order.
 *
 * When changedFields is provided, only transitively affected fields are
 * recomputed. Otherwise all derives are recomputed.
 *
 * @param {Object} graph - {topo_order: string[], deps: Object, dependents: Object}
 * @param {Object} fns - {name: (state) => value} compute functions
 * @param {Object} state - mutable state object (updated in place)
 * @param {string[]|null} changedFields - fields that changed, or null for all
 */
/**
 * Parse raw graph metadata from generated JS modules.
 *
 * @param {Object} rawGraph - raw graph from __graph__ property
 * @returns {{topo_order: string[], deps: Object, dependents: Object}}
 */
export function parseGraph(rawGraph) {
  if (!rawGraph || typeof rawGraph !== "object" || !Array.isArray(rawGraph.topo_order)) {
    return { topo_order: [], deps: {}, dependents: {} };
  }
  return rawGraph;
}

/**
 * Add a leaf derive to the graph (no other derives depend on it).
 *
 * Mutates the graph in place — adds to deps, dependents, and topo_order.
 *
 * @param {Object} graph - {topo_order, deps, dependents}
 * @param {string} name - derive field name
 * @param {string[]} fieldDeps - fields this derive depends on
 */
export function addLeafDerive(graph, name, fieldDeps) {
  if (graph.deps[name]) return; // already present

  graph.deps[name] = fieldDeps;
  for (const dep of fieldDeps) {
    if (!graph.dependents[dep]) graph.dependents[dep] = [];
    if (!graph.dependents[dep].includes(name)) graph.dependents[dep].push(name);
  }
  if (!graph.topo_order.includes(name)) graph.topo_order.push(name);
}

/**
 * Recompute derived fields in topological order.
 *
 * When changedFields is provided, only transitively affected fields are
 * recomputed. Otherwise all derives are recomputed.
 *
 * @param {Object} graph - {topo_order: string[], deps: Object, dependents: Object}
 * @param {Object} fns - {name: (state) => value} compute functions
 * @param {Object} state - mutable state object (updated in place)
 * @param {string[]|null} changedFields - fields that changed, or null for all
 */
export function recomputeGraph(graph, fns, state, changedFields = null) {
  let toRecompute;

  if (!changedFields) {
    toRecompute = graph.topo_order;
  } else {
    const affected = findAffected(graph.dependents, changedFields);
    toRecompute = graph.topo_order.filter(f => affected.has(f));
  }

  for (const name of toRecompute) {
    const fn = fns[name];
    if (fn) {
      try {
        state[name] = fn(state);
      } catch (err) {
        if (typeof console !== "undefined" && console.debug) {
          console.debug(`[Lavash] Error computing derive ${name}:`, err.message);
        }
      }
    }
  }
}
